task/task.go

```diff
@@ -0,0 +1,246 @@
+	// is NOT an error, but we really shouldn't try signalling
+	// it again; so just return
+	if t.statusCode != api.TaskStatusCode_Running {
+		return nil
+	}
+
+	// Signal the task to quit
+	t.statusCode = api.TaskStatusCode_Signalled
+	err := t.cmd.Process.Signal(syscall.SIGTERM)
+	if err != nil {
+		return err
+	}
+
+	// Schedule a timeout to kill the task if it refuses to exit
+	// after being asked politely. Will exit without killing the
+	// process if the task's `done` channel is closed before the
+	// timeout expires.
+	go runBrutalKillTimeout(t, timeout)
+
+	return nil
+}
+
+func runBrutalKillTimeout(t *Task, timeout time.Duration) {
+	select {
+	case <-time.After(timeout):
+		// The timeout has expired, and now it's unfortunately time
+		// to kill the process brutally
+		t.brutalKill()
+
+	case <-t.Done():
+		// The task has exited naturally before the timeout expired. No
+		// need to do anything other than let this goroutine exit
+		// naturally
+		return
+	}
+}
+
+// Done returns a channel that will be closed when the task
+// completed (for whatever reason), in the style of
+// context.Context.
+func (t *Task) Done() <-chan struct{} {
+	return t.done
+}
+
+// brutalKill kills the underlying process, giving no chance for cleanup.
+func (t *Task) brutalKill() {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	t.statusCode = api.TaskStatusCode_BrutallyKilled
+	err := t.cmd.Process.Kill()
+	if err != nil {
+		// Seems a bit excessive to panic here; The process just may have
+		// exited between us deciding to kill it and us actually doing it.
+		// For now we will just log the error so we have some visibility
+		// and move on.
+		log.Printf("Failed to kill task: %s", err.Error())
+	}
+}
+
+func cloneSlice(src []byte) []byte {
+	result := make([]byte, len(src))
+	copy(result, src)
+	return result
+}
+
+// Stdout creates and returns a copy of the current stdout data.
+func (t *Task) Stdout() []byte {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+	return cloneSlice(t.stdout.Bytes())
+}
+
+// Stderr creates and returns a copy of the current stderr data.
+func (t *Task) Stderr() []byte {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+	return cloneSlice(t.stderr.Bytes())
+}
+
+// monitor is executed in a goroutine and moves the process exit code (in the
+// form of an *os.Process) into place when the underlying process has exited
+func (t *Task) monitor() error {
+	// Wait for the underying process to complete before we try and force all
+	// of the IO streams to be flushed and closed, otherwise we have a data
+	// race on the Cmd that we wrap.
+
+	var exitCode int
+	switch err := t.cmd.Wait().(type) {
+	case nil:
+		exitCode = 0
+
+	case *exec.ExitError:
+		// For our purposes, this counts as a clean exit
+		exitCode = err.ProcessState.ExitCode()
+
+	default:
+		return err
+	}
+
+	// Now that we *know* the underlying process has finished, we can clean
+	// up the Cmd while we have it locked, averting the data race
```

Comment:
Looks like there's still some racy-ness going on as the task can still be marked as killed after having finished?

Reply:
I've added a check to the killing code that will only blow away a task in the `Signalled` state, which should prevent the signal timeout handler from killing an already-dead process.